File: test-stack.js
#!/usr/bin/env node
const http = require('http');

const checks = [
  { name: 'neo4j', host: 'localhost', port: 7474, path: '/', allowStatuses: [301, 200, 404] },
  { name: 'grafana', host: 'localhost', port: 3001, path: '/api/health', allowStatuses: [200] },
  { name: 'qdrant', host: 'localhost', port: 6333, path: '/', allowStatuses: [200, 404, 302] },
];

function checkHttp({ name, host, port, path, allowStatuses = [] }) {
  return new Promise((resolve) => {
    const start = Date.now();
    const req = http.request({ host, port, path, method: 'GET', timeout: 5000 });
    req.on('error', () => resolve({ name, ok: false, status: 0, duration: Date.now() - start }));
    req.on('timeout', () => {
      req.destroy();
      resolve({ name, ok: false, status: 0, duration: Date.now() - start });
    });
    req.on('response', (res) => {
      const duration = Date.now() - start;
      const ok = allowStatuses.includes(res.statusCode);
      resolve({ name, ok, status: res.statusCode, duration });
    });
  });
}

async function main() {
  const checksToRun = checks.map(checkHttp);
  const results = await Promise.all(checksToRun);
  let failed = 0;
  for (const r of results) {
    const status = r.ok ? 'PASS' : 'FAIL';
    console.log(`[${status}] ${r.name}: ${r.status} (${r.duration}ms)`);
    if (!r.ok) failed++;
  }

  if (failed) {
    console.error(`\n${failed} check(s) failed.`);
    process.exit(1);
  }

  console.log('\nStack health checks passed.');
}

main();

File: test_stack.py
#!/usr/bin/env python3
"""Smoke test: verify Docker services are running."""

import socket
import sys


def check_tcp(host, port, timeout=5):
    """Check if a TCP port is open."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(timeout)
    result = sock.connect_ex((host, port))
    sock.close()
    return result == 0


def main():
    services = [
        ("neo4j", 7474),
        ("grafana", 3001),
        ("postgres", 5432),
        ("redis", 6379),
        ("qdrant", 6333),
    ]

    all_ok = True
    for name, port in services:
        if check_tcp("localhost", port):
            print(f"OK {name}: port {port} open")
        else:
            print(f"FAIL {name}: port {port} closed")
            all_ok = False

    if all_ok:
        print("\nAll services are running!")
        sys.exit(0)
    else:
        print("\nSome services are not running!")
        sys.exit(1)


if __name__ == "__main__":
    main()
